assets/app/scripts/app.js

```diff
@@ -351,3 +351,103 @@ angular
+  var api = {
+    k8s: {},
+    openshift: {}
+  };
+  var apis = {};
+  var API_DISCOVERY_ERRORS = [];
+  var protocol = window.location.protocol + "//";
+ 
+ // Fetch /api/v1 for legacy k8s resources, we will never bump the version of these legacy apis so fetch version immediately
+  var k8sBaseURL = protocol + window.OPENSHIFT_CONFIG.api.k8s.hostPort + window.OPENSHIFT_CONFIG.api.k8s.prefix;
+  var k8sDeferred = $.get(k8sBaseURL + "/v1")
+  .done(function(data) {
+    var resources = {};
+    _.each(data.resources, function(resource) {
+      resources[resource.name] = resource;
+    });
+    api.k8s.v1 = resources;
+  })
+  .fail(function(data, textStatus, jqXHR) {
+    API_DISCOVERY_ERRORS.push({
+      data: data,
+      textStatus: textStatus,
+      xhr: jqXHR
+    });
+  });
+  
+  // Fetch /oapi/v1 for legacy openshift resources, we will never bump the version of these legacy apis so fetch version immediately
+  var osBaseURL = protocol + window.OPENSHIFT_CONFIG.api.openshift.hostPort + window.OPENSHIFT_CONFIG.api.openshift.prefix;
+  var osDeferred = $.get(osBaseURL + "/v1")
+  .done(function(data) {
+    var resources = {};
+    _.each(data.resources, function(resource) {
+      resources[resource.name] = resource;
+    });
+    api.openshift.v1 = resources;
+  })
+  .fail(function(data, textStatus, jqXHR) {
+    API_DISCOVERY_ERRORS.push({
+      data: data,
+      textStatus: textStatus,
+      xhr: jqXHR
+    });
+  });
+  
+  // Fetch /apis to get the list of groups and versions, then fetch each group/version
+  var apisDeferred = $.Deferred();
+  var apisBaseURL = protocol + window.OPENSHIFT_CONFIG.apis.hostPort + window.OPENSHIFT_CONFIG.apis.prefix;
+  $.get(apisBaseURL)
+  .done(function(data) {
+    var apisDeferredVersions = [];
+    _.each(data.groups, function(group) {
+      apis[group.name] = group;
+      var versions = {};
+      _.each(group.versions, function(version) {
+        versions[version.version] = version;
+        apisDeferredVersions.push($.get(apisBaseURL + "/" + version.groupVersion)
+        .done(function(data) {
+          var resources = {};
+          _.each(data.resources, function(resource) {
+            resources[resource.name] = resource;
+          });         
+          version.resources = resources;
+        })
+        .fail(function(data, textStatus, jqXHR) {
+          API_DISCOVERY_ERRORS.push({
+            data: data,
+            textStatus: textStatus,
+            xhr: jqXHR
+          });
+        }));
+      });
+      group.versions = versions;
+    });
+    $.when.apply(this, apisDeferredVersions).then(function() {
+      apisDeferred.resolve();
+    });    
+  })
+  .fail(function(data, textStatus, jqXHR) {
+    API_DISCOVERY_ERRORS.push({
+      data: data,
+      textStatus: textStatus,
+      xhr: jqXHR
+    });
+    apisDeferred.reject();
+  });
+  
+  $.when(k8sDeferred,osDeferred,apisDeferred).then(function() {
```

Comment:
apisDeferred is the only synthetic one, and that one we have to wait for
all nested calls to finish so I'm not sure how you do that another way?
On Apr 18, 2016 5:08 PM, "Jordan Liggitt" <notifications@github.com> wrote:

> In assets/app/scripts/app.js
> <https://github.com/jwforres/origin/pull/38#discussion_r60134450>:
>
> > +      group.versions = versions;
> > +    });
> > +    $.when.apply(this, apisDeferredVersions).then(function() {
> > +      apisDeferred.resolve();
> > +    });
> > +  })
> > +  .fail(function(data, textStatus, jqXHR) {
> > +    API_DISCOVERY_ERRORS.push({
> > +      data: data,
> > +      textStatus: textStatus,
> > +      xhr: jqXHR
> > +    });
> > +    apisDeferred.reject();
> > +  });
> > +
> > +  $.when(k8sDeferred,osDeferred,apisDeferred).then(function() {
>
> can you use the promises returned from $.get() instead of synthetic ones?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/jwforres/origin/pull/38/files/af570740dc6ecf32ce69439fbd08b88cc600794f#r60134450>
>